File: java/0014-longest-common-prefix.java
import java.util.*;

/**
 * LeetCode Problem 0014: Longest Common Prefix
 *
 * Problem:
 * Write a function to find the longest common prefix string amongst an array of strings.
 * If there is no common prefix, return an empty string "".
 *
 * Examples:
 * 1. Input: strs = ["flower","flow","flight"], Output: "fl"
 * 2. Input: strs = ["dog","racecar","car"], Output: ""
 */
public class LongestCommonPrefix {

    /**
     * Q1: Horizontal scanning - compare first string with all others
     *
     * Time Complexity: O(n * m) where n is number of strings, m is min length
     * Space Complexity: O(1)
     */
    public static String longestCommonPrefixHorizontal(String[] strs) {
        if (strs == null || strs.length == 0) return "";

        String prefix = strs[0];

        for (int i = 1; i < strs.length; i++) {
            while (!strs[i].startsWith(prefix)) {
                prefix = prefix.substring(0, prefix.length() - 1);
                if (prefix.isEmpty()) return "";
            }
        }

        return prefix;
    }

    /**
     * Q2: Vertical scanning - compare characters column by column
     *
     * Time Complexity: O(n * m)
     * Space Complexity: O(1)
     */
    public static String longestCommonPrefixVertical(String[] strs) {
        if (strs == null || strs.length == 0) return "";

        for (int i = 0; i < strs[0].length(); i++) {
            char c = strs[0].charAt(i);
            for (int j = 1; j < strs.length; j++) {
                if (i >= strs[j].length() || strs[j].charAt(i) != c) {
                    return strs[0].substring(0, i);
                }
            }
        }

        return strs[0];
    }

    /**
     * Q3: Divide and conquer
     *
     * Time Complexity: O(n * m)
     * Space Complexity: O(m) for recursion stack
     */
    public static String longestCommonPrefixDivideConquer(String[] strs) {
        if (strs == null || strs.length == 0) return "";
        return lcp(strs, 0, strs.length - 1);
    }

    private static String lcp(String[] strs, int left, int right) {
        if (left == right) return strs[left];

        int mid = left + (right - left) / 2;
        String lcpLeft = lcp(strs, left, mid);
        String lcpRight = lcp(strs, mid + 1, right);

        return common(lcpLeft, lcpRight);
    }

    private static String common(String left, String right) {
        int minLen = Math.min(left.length(), right.length());
        for (int i = 0; i < minLen; i++) {
            if (left.charAt(i) != right.charAt(i)) {
                return left.substring(0, i);
            }
        }
        return left.substring(0, minLen);
    }

    /**
     * Q4: Binary search on length
     *
     * Time Complexity: O(n * m * log(m))
     * Space Complexity: O(1)
     */
    public static String longestCommonPrefixBinarySearch(String[] strs) {
        if (strs == null || strs.length == 0) return "";

        int minLen = Integer.MAX_VALUE;
        for (String s : strs) {
            minLen = Math.min(minLen, s.length());
        }

        int left = 0, right = minLen;

        while (left < right) {
            int mid = left + (right - left + 1) / 2;
            if (isCommonPrefix(strs, mid)) {
                left = mid;
            } else {
                right = mid - 1;
            }
        }

        return strs[0].substring(0, left);
    }

    private static boolean isCommonPrefix(String[] strs, int length) {
        String prefix = strs[0].substring(0, length);
        for (int i = 1; i < strs.length; i++) {
            if (!strs[i].startsWith(prefix)) {
                return false;
            }
        }
        return true;
    }

    /**
     * Q5: Character-by-character with early termination
     *
     * Time Complexity: O(n * m)
     * Space Complexity: O(1)
     */
    public static String longestCommonPrefixCharByChar(String[] strs) {
        if (strs == null || strs.length == 0) return "";
        if (strs.length == 1) return strs[0];

        int minLen = Integer.MAX_VALUE;
        for (String s : strs) {
            minLen = Math.min(minLen, s.length());
        }

        for (int i = 0; i < minLen; i++) {
            char c = strs[0].charAt(i);
            for (int j = 1; j < strs.length; j++) {
                if (strs[j].charAt(i) != c) {
                    return strs[0].substring(0, i);
                }
            }
        }

        return strs[0].substring(0, minLen);
    }

    /**
     * Q6: Trie-based approach
     *
     * Time Complexity: O(n * m)
     * Space Complexity: O(n * m)
     */
    public static String longestCommonPrefixTrie(String[] strs) {
        if (strs == null || strs.length == 0) return "";

        class TrieNode {
            Map<Character, TrieNode> children = new HashMap<>();
            int count = 0;
        }

        TrieNode root = new TrieNode();

        // Build trie
        for (String word : strs) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                node.children.putIfAbsent(c, new TrieNode());
                node = node.children.get(c);
                node.count++;
            }
        }

        // Find LCP
        StringBuilder result = new StringBuilder();
        TrieNode node = root;
        while (node.children.size() == 1) {
            char c = node.children.keySet().iterator().next();
            TrieNode child = node.children.get(c);
            if (child.count == strs.length) {
                result.append(c);
                node = child;
            } else {
                break;
            }
        }

        return result.toString();
    }

    /**
     * Q7: Using StringBuilder with comparison
     *
     * Time Complexity: O(n * m)
     * Space Complexity: O(m)
     */
    public static String longestCommonPrefixStringBuilder(String[] strs) {
        if (strs == null || strs.length == 0) return "";

        StringBuilder prefix = new StringBuilder();

        for (int i = 0; i < strs[0].length(); i++) {
            char c = strs[0].charAt(i);
            for (int j = 1; j < strs.length; j++) {
                if (i >= strs[j].length() || strs[j].charAt(i) != c) {
                    return prefix.toString();
                }
            }
            prefix.append(c);
        }

        return prefix.toString();
    }

    /**
     * Q8: Two-pointer scanning
     *
     * Time Complexity: O(n * m)
     * Space Complexity: O(1)
     */
    public static String longestCommonPrefixTwoPointer(String[] strs) {
        if (strs == null || strs.length == 0) return "";

        int col = 0;
        while (col < strs[0].length()) {
            char c = strs[0].charAt(col);
            for (int row = 1; row < strs.length; row++) {
                if (col >= strs[row].length() || strs[row].charAt(col) != c) {
                    return strs[0].substring(0, col);
                }
            }
            col++;
        }

        return strs[0];
    }

    /**
     * Q9: Iterative with substring comparisons
     *
     * Time Complexity: O(n * m)
     * Space Complexity: O(1)
     */
    public static String longestCommonPrefixIterative(String[] strs) {
        if (strs == null || strs.length == 0) return "";

        for (int i = 0; i < strs[0].length(); i++) {
            for (int j = 1; j < strs.length; j++) {
                if (i >= strs[j].length() ||
                    strs[0].charAt(i) != strs[j].charAt(i)) {
                    return strs[0].substring(0, i);
                }
            }
        }

        return strs[0];
    }

    /**
     * Q10: Using stream API
     *
     * Time Complexity: O(n * m)
     * Space Complexity: O(1)
     */
    public static String longestCommonPrefixStream(String[] strs) {
        if (strs == null || strs.length == 0) return "";

        for (int i = 0; i < strs[0].length(); i++) {
            final int index = i;
            char c = strs[0].charAt(index);

            boolean allMatch = Arrays.stream(strs, 1, strs.length)
                    .allMatch(s -> index < s.length() && s.charAt(index) == c);

            if (!allMatch) {
                return strs[0].substring(0, index);
            }
        }

        return strs[0];
    }

    // Test cases and main function
    public static void main(String[] args) {
        String[][] testCases = {
                {"flower", "flow", "flight"},
                {"dog", "racecar", "car"},
                {"ab"},
                {"a"},
                {"interspecies", "interstellar", "interstate"},
                {"abc"},
                {"abab", "aba", "abc"},
                {"a", "a", "a", "a"},
        };
        String[] expected = {"fl", "", "ab", "a", "inters", "abc", "ab", "a"};

        java.lang.reflect.Method[] methods = LongestCommonPrefix.class.getDeclaredMethods();
        List<java.lang.reflect.Method> testMethods = new ArrayList<>();

        for (java.lang.reflect.Method method : methods) {
            if (method.getName().startsWith("longestCommonPrefix") &&
                method.getReturnType() == String.class) {
                testMethods.add(method);
            }
        }

        Collections.sort(testMethods, Comparator.comparing(java.lang.reflect.Method::getName));

        int questionNum = 1;
        for (java.lang.reflect.Method method : testMethods) {
            System.out.println("\nQ" + questionNum + ": " + method.getName());
            System.out.println("-".repeat(60));

            boolean allPassed = true;
            for (int i = 0; i < testCases.length; i++) {
                try {
                    String result = (String) method.invoke(null, (Object) testCases[i]);
                    boolean passed = result.equals(expected[i]);
                    allPassed = allPassed && passed;
                    String status = passed ? "✓" : "✗";
                    System.out.println("  " + status + " longestCommonPrefix(" +
                        Arrays.toString(testCases[i]) + ") = \"" + result +
                        "\" (expected: \"" + expected[i] + "\")");
                } catch (Exception e) {
                    System.out.println("  ✗ Error: " + e.getMessage());
                    allPassed = false;
                }
            }

            if (allPassed) {
                System.out.println("  All tests passed!");
            }

            questionNum++;
        }
    }
}

File: python/0014-longest-common-prefix.py
"""
LeetCode Problem 0014: Longest Common Prefix

Problem:
Write a function to find the longest common prefix string amongst an array of strings.
If there is no common prefix, return an empty string "".

Examples:
1. Input: strs = ["flower","flow","flight"], Output: "fl"
2. Input: strs = ["dog","racecar","car"], Output: ""
3. Input: strs = ["ab"], Output: "ab"
4. Input: strs = ["a"], Output: "a"

Constraints:
- 1 <= strs.length <= 200
- 0 <= strs[i].length <= 200
- strs[i] consists of only lowercase English letters
"""

# Q1: Horizontal scanning - compare first string with all others
def longestCommonPrefix_horizontal(strs: list[str]) -> str:
    """
    Time Complexity: O(n * m) where n is number of strings, m is min length
    Space Complexity: O(1)

    Approach: Start with first string as prefix, then trim it char by char
    until it matches prefix of all strings.
    """
    if not strs:
        return ""

    prefix = strs[0]

    for i in range(1, len(strs)):
        while not strs[i].startswith(prefix):
            prefix = prefix[:-1]
            if not prefix:
                return ""

    return prefix


# Q2: Vertical scanning - compare characters column by column
def longestCommonPrefix_vertical(strs: list[str]) -> str:
    """
    Time Complexity: O(n * m) where n is number of strings, m is min length
    Space Complexity: O(1)

    Approach: Compare character by character at each position across all strings.
    """
    if not strs:
        return ""

    for i in range(len(strs[0])):
        char = strs[0][i]
        for j in range(1, len(strs)):
            if i >= len(strs[j]) or strs[j][i] != char:
                return strs[0][:i]

    return strs[0]


# Q3: Divide and conquer
def longestCommonPrefix_divide_conquer(strs: list[str]) -> str:
    """
    Time Complexity: O(n * m)
    Space Complexity: O(m) for recursion stack

    Approach: Divide array in half, find LCP of each half, then merge results.
    """
    if not strs:
        return ""

    def lcp(strs, left, right):
        if left == right:
            return strs[left]

        mid = (left + right) // 2
        lcp_left = lcp(strs, left, mid)
        lcp_right = lcp(strs, mid + 1, right)

        return common(lcp_left, lcp_right)

    def common(left, right):
        min_len = min(len(left), len(right))
        for i in range(min_len):
            if left[i] != right[i]:
                return left[:i]
        return left[:min_len]

    return lcp(strs, 0, len(strs) - 1)


# Q4: Binary search on length
def longestCommonPrefix_binary_search(strs: list[str]) -> str:
    """
    Time Complexity: O(n * m * log(m)) where n is number of strings, m is min length
    Space Complexity: O(1)

    Approach: Binary search on the length of the prefix.
    """
    if not strs:
        return ""

    def is_common_prefix(length):
        prefix = strs[0][:length]
        for i in range(1, len(strs)):
            if not strs[i].startswith(prefix):
                return False
        return True

    left, right = 0, min(len(s) for s in strs)

    while left < right:
        mid = (left + right + 1) // 2
        if is_common_prefix(mid):
            left = mid
        else:
            right = mid - 1

    return strs[0][:left]


# Q5: Trie-based approach
def longestCommonPrefix_trie(strs: list[str]) -> str:
    """
    Time Complexity: O(n * m) to build trie, O(m) to find LCP
    Space Complexity: O(n * m) for trie storage

    Approach: Build a trie of all strings, then traverse down the trie
    while there's only one child at each node.
    """
    if not strs:
        return ""

    class TrieNode:
        def __init__(self):
            self.children = {}
            self.count = 0

    root = TrieNode()

    # Build trie
    for word in strs:
        node = root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
            node.count += 1

    # Find LCP
    result = ""
    node = root
    while node.children and len(node.children) == 1:
        char = list(node.children.keys())[0]
        if node.children[char].count == len(strs):
            result += char
            node = node.children[char]
        else:
            break

    return result


# Q6: Character-by-character with early termination
def longestCommonPrefix_char_by_char(strs: list[str]) -> str:
    """
    Time Complexity: O(n * m)
    Space Complexity: O(1)

    Approach: Check each position until a mismatch is found.
    """
    if not strs:
        return ""

    if len(strs) == 1:
        return strs[0]

    min_len = min(len(s) for s in strs)

    for i in range(min_len):
        char = strs[0][i]
        for s in strs[1:]:
            if s[i] != char:
                return strs[0][:i]

    return strs[0][:min_len]


# Q7: Using zip and set
def longestCommonPrefix_zip(strs: list[str]) -> str:
    """
    Time Complexity: O(n * m)
    Space Complexity: O(1)

    Approach: Use zip to group characters by position, check for consistency.
    """
    if not strs:
        return ""

    result = ""
    for chars in zip(*strs):
        if len(set(chars)) == 1:
            result += chars[0]
        else:
            break

    return result


# Q8: Recursive horizontal scanning
def longestCommonPrefix_recursive(strs: list[str]) -> str:
    """
    Time Complexity: O(n * m)
    Space Complexity: O(m) for recursion stack

    Approach: Recursively compare strings.
    """
    if not strs:
        return ""

    def helper(strs, index):
        if index >= len(strs[0]):
            return strs[0][:index]

        char = strs[0][index]
        for s in strs[1:]:
            if index >= len(s) or s[index] != char:
                return strs[0][:index]

        return helper(strs, index + 1)

    return helper(strs, 0)


# Q9: Generator-based approach
def longestCommonPrefix_generator(strs: list[str]) -> str:
    """
    Time Complexity: O(n * m)
    Space Complexity: O(1)

    Approach: Use generator to lazily evaluate prefixes.
    """
    if not strs:
        return ""

    def prefix_gen():
        for i in range(len(strs[0])):
            char = strs[0][i]
            if all(i < len(s) and s[i] == char for s in strs[1:]):
                yield char
            else:
                break

    return "".join(prefix_gen())


# Q10: Slicing with all() function
def longestCommonPrefix_slicing(strs: list[str]) -> str:
    """
    Time Complexity: O(n * m)
    Space Complexity: O(1)

    Approach: For each position, slice all strings and check consistency.
    """
    if not strs:
        return ""

    result = ""
    for i in range(len(strs[0])):
        chars = [s[i] if i < len(s) else None for s in strs]
        if len(set(chars)) == 1 and chars[0] is not None:
            result += chars[0]
        else:
            break

    return result


if __name__ == "__main__":
    test_cases = [
        (["flower", "flow", "flight"], "fl"),
        (["dog", "racecar", "car"], ""),
        (["ab"], "ab"),
        (["a"], "a"),
        (["interspecies", "interstellar", "interstate"], "inters"),
        (["abc"], "abc"),
        (["abab", "aba", "abc"], "ab"),
        (["", "b"], ""),
        (["a", "a", "a", "a"], "a"),
        (["aca", "cba", "bda"], ""),
    ]

    functions = [
        longestCommonPrefix_horizontal,
        longestCommonPrefix_vertical,
        longestCommonPrefix_divide_conquer,
        longestCommonPrefix_binary_search,
        longestCommonPrefix_trie,
        longestCommonPrefix_char_by_char,
        longestCommonPrefix_zip,
        longestCommonPrefix_recursive,
        longestCommonPrefix_generator,
        longestCommonPrefix_slicing,
    ]

    for i, func in enumerate(functions, 1):
        print(f"\nQ{i}: {func.__name__}")
        print("-" * 60)
        all_passed = True
        for strs, expected in test_cases:
            result = func(strs)
            passed = result == expected
            all_passed = all_passed and passed
            status = "✓" if passed else "✗"
            print(f"  {status} longestCommonPrefix({strs}) = '{result}' (expected: '{expected}')")

        if all_passed:
            print(f"  All tests passed!")
